doc/source/data/performance-tips.rst

```diff
@@ -25,17 +25,122 @@ Tuning read parallelism
-3. The in-memory data size is estimated. If the parallelism would create in-memory blocks that are larger on average than the target block size (512MiB), the parallelism is increased until the blocks are < 512MiB in size.
+The ``parallelism`` parameter passed to Ray Data's :ref:`read APIs <input-output>` specifies the number of read tasks that should be used.
+Usually, if the read is followed by a :func:`~ray.data.Dataset.map` or :func:`~ray.data.Dataset.map_batches`, the map is fused with the read; therefore ``parallelism`` also determines the parallelism used during map.
 
-Occasionally, it's advantageous to manually tune the parallelism to optimize the application. You can do this by setting the ``parallelism`` parameter.
-For example, use ``ray.data.read_parquet(path, parallelism=1000)`` to force up to 1000 read tasks to be created.
+The default value for ``parallelism`` is decided based on the following heuristics, applied in order:
+
+1. We start with the default parallelism of 200. This can be overridden by setting :class:`DataContext.min_parallelism <ray.data.context.DataContext>`.
+2. Min block size (default=1MiB). If the parallelism would make blocks smaller than this threshold, the parallelism is reduced to avoid the overhead of tiny blocks. This can be overridden by setting :class:`DataContext.target_min_block_size <ray.data.context.DataContext>` (bytes).
+3. Max block size (default=128MiB). If the parallelism would make blocks larger than this threshold, the parallelism is increased to avoid OOMs during processing. This can be overridden by setting :class:`DataContext.target_max_block_size <ray.data.context.DataContext>` (bytes).
+4. Available CPUs. If the parallelism cannot make use of all the available CPUs in the cluster, the parallelism is increased until it can. Ray Data chooses read tasks be at least 2x the number of available CPUs.
+
+Occasionally, it's advantageous to manually tune the parallelism to optimize the application.
+For example, the following code will batch multiple files into the same read task to avoid creating blocks that are too large.
+
+.. testcode::
+
+    import ray
+    # Pretend we have two CPUs.
+    ray.init(num_cpus=2)
+
+    # Repeat the iris.csv file 16 times.
+    ds = ray.data.read_images(["example://iris.csv"] * 16)
+    print(ds.materialize())
+    # 2023-11-20 14:28:47,597 INFO plan.py:760 -- Using autodetected parallelism=4 for stage ReadCSV to satisfy parallelism at least twice the available number of CPUs (2).
+    # MaterializedDataset(
+    #    num_blocks=4,
+    #    num_rows=2400,
+    #    ...
+    # )
+
+But suppose that you knew that you wanted to read all 16 files in parallel.
+This could, for example, because you know that additional CPUs should get added to the cluster by the autoscaler or because you want the downstream stage to transform each file's contents in parallel.
```

Comment:
Missing word?
```suggestion
This could be, for example, because you know that additional CPUs should get added to the cluster by the autoscaler or because you want the downstream stage to transform each file's contents in parallel.
```